arch/esp32s3: lcd add fb pandisplay

liamHowatt opened the pull request:
cc @XuNeo @FASTSHIFT

## Summary

Make the esp32s3 LCD FB device support double buffering. Leverage the FB `pandisplay` callback for users to achieve [double buffering in applications such as LVGL](https://github.com/lvgl/lvgl/blob/f89ac3677e6666449835d765d182f5515770c984/src/drivers/nuttx/lv_nuttx_fbdev.c#L275-L288). Use DMA interrupt to facilitate vsync.

## Impact

I would like feedback on how this change should be managed. I made it so that the feature will be active if there is 1 LCD layer. The impact is that double the memory will be allocated compared to before this change, which could cause users to have insufficient memory after upgrading. Should it be a Kconfig option?

## Testing

Mainboard: ESP32-S3-LCD-EV-Board-MB_V1.5
LCD Subboard: ESP32-S3-LCD-EV-Board-SUB2_V1.5

config: `esp32s3-lcd-ev:lvgl`

For the specific board model:
```
kconfig-tweak --set-val CONFIG_ESP32S3_I2C0_SCLPIN 48
kconfig-tweak --set-val CONFIG_ESP32S3_I2C0_SDAPIN 47
kconfig-tweak --set-val CONFIG_ESP32S3_LCD_DATA6_PIN 8
kconfig-tweak --set-val CONFIG_ESP32S3_LCD_DATA7_PIN 18
yes '' | make oldconfig
```

LVGL will [detect the effect of the changes](https://github.com/lvgl/lvgl/blob/f89ac3677e6666449835d765d182f5515770c984/src/drivers/nuttx/lv_nuttx_fbdev.c#L145-L154) in the PR and double buffering will take effect without any other changes.

nuttxpr commented:
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



This PR appears to mostly meet the NuttX requirements, but has some gaps that need addressing before it's ready. Here's a breakdown:

**Strengths:**

* **Clear Summary:** The summary explains the "why," "what," and "how" of the change effectively.  The link to the LVGL code showing the usage is helpful.
* **Impact Acknowledged:** The impact section correctly identifies the potential memory issue and seeks feedback on how to handle it. This proactive approach is good.
* **Testing Information Provided:**  The testing section specifies the hardware and software environment used for verification. The instructions for configuring the board are clear.  Referencing how LVGL detects the changes is also good.


**Weaknesses:**

* **Missing Issue References:**  The summary should link to a related NuttX or NuttX Apps issue if one exists. If not, consider creating one to track the feature request/bug this PR addresses.
* **Impact Needs More Detail:** While the memory impact is mentioned, other impacts need to be addressed more thoroughly.  Even if the answer is "NO", a brief explanation is helpful for reviewers.  For example:
    * **Impact on user:**  Even if LVGL adapts automatically, will *other* users of the framebuffer device need to change their code?
    * **Impact on build:**  Does adding double buffering change any build dependencies or configurations?
    * **Impact on hardware:** Is this change specific to the ESP32-S3, or could it be applied to other LCD drivers?
    * **Impact on documentation:** Does this change require updating any documentation (e.g., the framebuffer driver documentation, board configuration documentation)?  If so, has the documentation been updated in the PR?
    * **Impact on security:** While unlikely, it's good practice to consider security implications of any change.  Explicitly stating "NO" with a brief justification is sufficient.
    * **Impact on compatibility:**  Does this change break compatibility with any existing applications or configurations?
* **Missing Testing Logs:** The testing section *describes* the testing but doesn't include any actual logs.  Include relevant log output before and after the change to demonstrate the functionality and the absence of regressions.  Demonstrate both single buffering and double buffering behavior.  Show what happens if the system has insufficient memory.
* **Kconfig Decision:** The PR raises the question about a Kconfig option but doesn't propose a solution.  Suggest a concrete Kconfig option (name, description, default value) in the PR description.  This allows reviewers to discuss the specific proposal.


**Recommendations:**

1. **Create a related issue:** File an issue in the appropriate repository (NuttX or NuttX-Apps) and link it in the PR summary.
2. **Complete the Impact section:** Address all impact points with specific details, even if the answer is "NO."
3. **Provide testing logs:** Include actual log output demonstrating the behavior before and after the change, covering both positive and negative cases (sufficient and insufficient memory).
4. **Propose a Kconfig option:**  Suggest a specific Kconfig option (name, description, default value) to control double buffering. Justify your choice (e.g., why enabled by default or disabled by default).


By addressing these points, the PR will be much stronger and easier for reviewers to evaluate.

tmedicci commented:
> I would like feedback on how this change should be managed. I made it so that the feature will be active if there is 1 LCD layer. The impact is that double the memory will be allocated compared to before this change, which could cause users to have insufficient memory after upgrading. Should it be a Kconfig option?

Yes, I think so. You can turn `ESP32S3_LCD_DBLBUFFERED` a Kconfig option that depends on CONFIG_ESP32S3_LCD_BUFFER_LAYERS being `= 1` (just like [here](https://github.com/apache/incubator-nuttx/blob/b6d2dc6c5487ffa957eaa524e02e9699b1f6dacc/drivers/lcd/Kconfig#L1279)) and can be enabled/disabled by the user (default to `y`, I think). You can insert in `help` the warning about the memory usage.

FASTSHIFT commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -623,6 +663,8 @@ static int IRAM_ATTR lcd_interrupt(int irq, void *context, void *arg)
       regval |= LCD_CAM_LCD_START_M;
       esp32s3_lcd_putreg(LCD_CAM_LCD_USER_REG, regval);
+
+      fb_remove_paninfo(&g_base_vtable, FB_NO_OVERLAY);
```
Add a return value check.

liamHowatt replied:
Take a look what I did in my latest changes. It's similar to the other users of this function.

xiaoxiang781216 commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -548,7 +564,31 @@ static int esp32s3_lcd_base_updatearea(struct fb_vtable_s *vtable,
+static int esp32s3_lcd_base_pandisplay(FAR struct fb_vtable_s *vtable,
+                                       FAR struct fb_planeinfo_s *pinfo)
+{
+  struct esp32s3_lcd_s *priv = &g_lcd_priv;
```
arch/board code could drop FAR, only the common code need it

liamHowatt commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -300,13 +310,16 @@ static const struct fb_videoinfo_s g_base_videoinfo =
 /* This structure provides the base layer interface */
 
-static const struct fb_vtable_s g_base_vtable =
+static struct fb_vtable_s g_base_vtable =
```
The `fb_remove_paninfo(&g_base_vtable, FB_NO_OVERLAY);` I added takes a `FAR struct fb_vtable_s *vtable`. Looks like `fb_remove_paninfo` does not modify it, though. Maybe the parameter could be made const? See https://github.com/apache/nuttx/blob/52b0df1ce6cdcc99b44700b20a2bc3351285eee0/drivers/video/fb.c#L1607

Otherwise, should I keep the vtable const in esp32s3_lcd.c and cast to non-`const`?

xiaoxiang781216 replied:
Ok, let's keep as it.

liamHowatt replied:
I understand why it has to be non-`const` now. There's a member `priv` at the end of the struct that gets assigned by fb.c. In `master` the `const` region is being written to.

liamHowatt commented:
I'm going to make some changes I think.

I did this some time ago and wanted to remember why I made it work for 1 layer only. I don't think there was a reason, after analyzing. However, I think the >1 layer support is incomplete. If there are two layers, `lcd_interrupt` alternates between layers, DMAing their framebuffers out. When the `getvideoinfo` or `getplaneinfo` ioctl is called, a random layer's info is returned. The layer which was last alternated to is the current layer and its info is returned. I'm going to update this PR to continue using the random `CURRENT_LAYER` getter pattern until it's properly addressed in another PR or even this PR if you wish.

The new `esp32s3_lcd_base_pandisplay` is blocking due to `esp32s3_dma_wait_idle`. I think I could/should use a critical section to set a flag which `lcd_interrupt` checks to know if the buffer should be panned and that's where `esp32s3_dma_setup` would be called.

Are force-pushes preferred? Squash merging isn't normally done?

xiaoxiang781216 commented:
Please update the patch inplace.

liamHowatt commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -548,7 +564,31 @@ static int esp32s3_lcd_base_updatearea(struct fb_vtable_s *vtable,
+
+  esp32s3_dma_wait_idle(priv->dma_channel, true);
+
+  esp32s3_dma_setup(layer->dmadesc,
```
Now it's in an ISR. Any advice for how to manage an error, considering the latest changes?

liamHowatt commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -548,7 +564,31 @@ static int esp32s3_lcd_base_updatearea(struct fb_vtable_s *vtable,
+
+  esp32s3_dma_wait_idle(priv->dma_channel, true);
+
+  esp32s3_dma_setup(layer->dmadesc,
```
It chains transfers together using a linked list of DMA descriptors. In this PR, the transfer size is not increasing, just the starting offset changes now. There's another place `esp32s3_dma_setup` is used in this file.

liamHowatt commented:
I've updated and validated it. It's much more correct now, despite the previous iteration technically working.

> can be enabled/disabled by the user (default to `y`, I think).

Okay, if you think it can be default. I've done it. I made it independent of layers, as mentioned in my previous comment.

@FASTSHIFT I added some checks. Please see my responses to your review.

xiaoxiang781216 commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -244,6 +258,11 @@ static int esp32s3_lcd_base_updatearea(struct fb_vtable_s *vtable,
 #endif
 
+#ifdef CONFIG_ESP32S3_LCD_DOUBLE_BUFFERED
+static int esp32s3_lcd_base_pandisplay(FAR struct fb_vtable_s *vtable,
```
remove all FAR

xiaoxiang781216 commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -603,10 +677,43 @@ static int IRAM_ATTR lcd_interrupt(int irq, void *context, void *arg)
+          fb_remove_paninfo(&g_base_vtable, FB_NO_OVERLAY);
+        }
+
+      union fb_paninfo_u info;
```
need move to the begin of function/block

FASTSHIFT commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -545,10 +569,59 @@ static int esp32s3_lcd_base_getplaneinfo(struct fb_vtable_s *vtable,
+{
+  if (pinfo->yoffset > ESP32S3_LCD_VRES_VIRTUAL - CONFIG_ESP32S3_LCD_VRES)
+    {
+      gerr("ERROR: pandisplay yoffset out of bounds\n");
```
print `pinfo->yoffset` value

FASTSHIFT commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -545,10 +569,59 @@ static int esp32s3_lcd_base_getplaneinfo(struct fb_vtable_s *vtable,
+  if (area->x + area->w > ESP32S3_LCD_HRES_VIRTUAL ||
+      area->y + area->h > ESP32S3_LCD_VRES_VIRTUAL)
+    {
+      gerr("ERROR: updatearea area is out of bounds\n");
```
print area info

xiaoxiang781216 commented on arch/xtensa/src/esp32s3/esp32s3_lcd.c:
```diff
@@ -545,10 +569,66 @@ static int esp32s3_lcd_base_getplaneinfo(struct fb_vtable_s *vtable,
+                         (area->y * ESP32S3_LCD_STRIDE +
+                          area->x * ESP32S3_LCD_DATA_WIDTH);
+
+  uint32_t size = (area->h - 1) * ESP32S3_LCD_STRIDE +
```
size and first_pixel need at the begin of function too to confirm c89 spec

jerpelea reviewed (approved):
please add a commit message
